client-tools/Shell/TelemetricsHandler.cpp

```diff
@@ -0,0 +1,380 @@
+    VPackBuilder& builder) {
+  std::string const originalUrl = "/test-redirect/redirect";
+  std::string url = originalUrl;
+  Result result = {TRI_ERROR_NO_ERROR};
+  uint32_t timeoutInSecs = 10;
+  uint32_t maxRedirects = 5;
+
+  uint32_t numRedirects = 0;
+
+  while (!_server.isStopping()) {
+    ClientManager clientManager(
+        _server.getFeature<HttpEndpointProvider, ClientFeature>(),
+        Logger::FIXME);
+    std::unique_lock lk(_mtx);
+
+    _httpClient = clientManager.getConnectedClient(true, false, false, 0);
+    if (_httpClient != nullptr && _httpClient->isConnected()) {
+      std::string const body = _telemetricsResult.toJson();
+      auto& clientParams = _httpClient->params();
+      clientParams.setRequestTimeout(30);
+      lk.unlock();
+
+      std::unordered_map<std::string, std::string> headers;
+      headers.emplace(StaticStrings::ContentTypeHeader,
+                      StaticStrings::MimeTypeJson);
+      headers.emplace(StaticStrings::ContentLength,
+                      std::to_string(body.size()));
+
+      std::unique_ptr<httpclient::SimpleHttpResult> response(
+          _httpClient->request(rest::RequestType::POST, url, body.data(),
+                               body.size(), headers));
+      lk.lock();
+      result = this->checkHttpResponse(response, false);
+      if (result.ok()) {
+        auto returnCode = response->getHttpReturnCode();
+        if (returnCode == 200) {
+          builder.add(response->getBodyVelocyPack()->slice());
+          break;
+        } else if (returnCode == 301 || returnCode == 302 ||
+                   returnCode == 307) {
+          if (numRedirects < maxRedirects) {
+            bool found;
+            url = response->getHeaderField(StaticStrings::Location, found);
+            if (found) {
+              numRedirects++;
+            } else {
+              url = originalUrl;
+              numRedirects = 0;
+            }
+          } else {
+            url = originalUrl;
+            numRedirects = 0;
+          }
+        }
+      } else {
+        builder.openObject();
+        builder.add("errorNum", VPackValue(_telemetricsResponse.errorNumber()));
+        builder.add("errorMessage",
+                    VPackValue(_telemetricsResponse.errorMessage()));
+        builder.close();
+        numRedirects = 0;
+      }
+    }
+
+    if (numRedirects == 0) {
+      timeoutInSecs *= 3;
+      timeoutInSecs = std::min<uint32_t>(600, timeoutInSecs);
+      _runCondition.wait_for(lk, std::chrono::seconds(timeoutInSecs),
+                             [this] { return _server.isStopping(); });
+    }
+  }
+}
+
+void TelemetricsHandler::sendTelemetricsToEndpoint() {
+  std::string const originalUrl =
+      "https://europe-west3-telemetrics-project.cloudfunctions.net/"
+      "telemetrics-cf-3";
+
+  std::string url = originalUrl;
+
+  Result result = {TRI_ERROR_NO_ERROR};
+  uint32_t timeoutInSecs = 10;
+  uint64_t sslProtocol = TLS_V12;
```

Comment:
Whenever possible, we should use TLS1.3. It has many advantages over TLS1.2 and should be understood by up-to-date servers.
```suggestion
  uint64_t sslProtocol = TLS_V13;
```